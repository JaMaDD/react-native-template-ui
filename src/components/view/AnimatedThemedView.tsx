import type { FC } from 'react';
import type { AnimatedThemedViewProps } from '../../types/view';
import AnimatedBox from './AnimatedBox';

const AnimatedThemedView: FC<AnimatedThemedViewProps> = ({
  backgroundColor = 'background',
  ...props
}) => {
  return (
    <AnimatedBox
      backgroundColor={backgroundColor}
      accessible={true}
      role={'group'}
      {...props}
    />
  );
};

export default AnimatedThemedView;
