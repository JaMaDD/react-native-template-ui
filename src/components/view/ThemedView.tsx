import type { FC } from 'react';
import type { ThemedViewProps } from '../../types/view';
import Box from './Box';

const ThemedView: FC<ThemedViewProps> = ({
  backgroundColor = 'background',
  ...props
}) => {
  return (
    <Box
      backgroundColor={backgroundColor}
      accessible={true}
      role={'group'}
      {...props}
    />
  );
};

export default ThemedView;
