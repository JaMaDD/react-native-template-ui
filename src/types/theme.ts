import type {
  BackgroundColorProps,
  BackgroundColorShorthandProps,
  BoxProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  SpacingProps,
  SpacingShorthandProps,
  TextProps,
} from '@shopify/restyle';

export type Theme = globalThis.ReactNativeTemplateTheme;

export type ThemeColors = keyof Theme['colors'];

export type ThemeSpacing = keyof Theme['spacing'];

export type ThemeBreakpoints = keyof Theme['breakpoints'];

export type ThemeZIndices = keyof Theme['zIndices'];

export type ThemeBorderRadii = keyof Theme['borderRadii'];

export type ThemeTextVariants = keyof Theme['textVariants'];

export type ThemeViewProps = BoxProps<Theme>;

export type ThemeTextProps = TextProps<Theme>;

export type ThemeColorProps = ColorProps<Theme>;

export type ThemeSpacingProps = SpacingProps<Theme> &
  SpacingShorthandProps<Theme>;

export type ThemeBackgroundColorProps = BackgroundColorProps<Theme> &
  BackgroundColorShorthandProps<Theme>;

export type ThemeBaseProps = ThemeSpacingProps &
  ThemeBackgroundColorProps &
  LayoutProps<Theme> &
  PositionProps<Theme>;
