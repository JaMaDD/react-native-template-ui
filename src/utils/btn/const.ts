export enum BtnScaleRatio {
  Square = 0.9,
  Rectangle = 0.98,
}

export enum OnPressDelayType {
  Throttle,
  Debounce,
}
