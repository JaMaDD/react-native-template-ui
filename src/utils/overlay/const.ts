import { IconSize } from '@jamadd/react-native-template-icons';
import { createContext } from 'react';
import type {
  ActionSheetContextVal,
  AlertContextVal,
} from '../../types/overlay';
import type { ThemeSpacing, ThemeTextVariants } from '../../types/theme';

/** @internal */
export const overlayMaxWidthPercent = '80%';

export enum OverlayDismissResultType {
  Toast = 'toast',
  Alert = 'alert',
  ActionSheet = 'actionSheet',
}

export const overlayDismissResultDefaultText = 'cancelByUser';

export enum ToastType {
  Top = 'top',
  TopLeft = 'topLeft',
  TopRight = 'topRight',
  Bottom = 'bottom',
  BottomLeft = 'bottomLeft',
  BottomRight = 'bottomRight',
}

export enum ToastDuration {
  Short = 3000,
  Long = 5000,
}

/** @internal */
export const AlertContext = createContext<AlertContextVal>({});

export enum AlertBtnType {
  Primary = 'primary',
  Secondary = 'secondary',
  Destructive = 'destructive',
}

/** @internal */
export const ActionSheetContext = createContext<Partial<ActionSheetContextVal>>(
  {}
);

/** @internal */
export const actionSheetOpenDuration = 300;

/** @internal */
export const actionSheetSnapDuration = 300;

/** @internal */
export const actionSheetDismissDuration = 150;

/** @internal */
export const actionSheetMaxHeightRatio = 0.6;

/** @internal */
export enum ActionSheetExpandableHeightRatio {
  Top = 0.9,
  Middle = 0.6,
  Bottom = 0.3,
}

/** @internal */
export const actionSheetHeaderIconSize: IconSize = IconSize.M;

/** @internal */
export const actionSheetHeaderPadding: ThemeSpacing = 'm';

/** @internal */
export const actionSheetOptListItemPadding: ThemeSpacing = 'm';

/** @internal */
export const actionSheetOptListItemTextVariant: ThemeTextVariants = 'textM';
