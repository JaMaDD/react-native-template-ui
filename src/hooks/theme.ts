import {
  useResponsiveProp,
  useTheme as useRestyleTheme,
  type PropValue,
  type ResponsiveValue,
} from '@shopify/restyle';
import { useColorScheme, type ColorSchemeName } from 'react-native';
import type { Theme, ThemeBreakpoints } from '../types/theme';
import { darkTheme, lightTheme } from '../utils/theme/const';

/** @internal */
export function useTheme() {
  return useRestyleTheme<Theme>();
}

export function useThemeColors() {
  return useTheme().colors;
}

export function useThemeSpacing() {
  return useTheme().spacing;
}

export function useThemeBreakpoints() {
  return useTheme().breakpoints;
}

export function useThemeBreakpointValues<T extends PropValue>(
  propValue: ResponsiveValue<T, Theme['breakpoints']>
) {
  return useResponsiveProp(propValue);
}

export function useThemeBreakpoint() {
  const responsiveVal: ResponsiveValue<PropValue, Theme['breakpoints']> = {
    smallPhone: 'smallPhone',
    phone: 'phone',
    smallTablet: 'smallTablet',
    tablet: 'tablet',
    desktop: 'desktop',
  };

  return useThemeBreakpointValues(responsiveVal) as ThemeBreakpoints;
}

export function useThemeZIndices() {
  return useTheme().zIndices;
}

export function useThemeBorderRadii() {
  return useTheme().borderRadii;
}

export function useThemeTextVariants() {
  return useTheme().textVariants;
}

export function useIsDarkColorScheme(overrideColorScheme?: ColorSchemeName) {
  const colorScheme = useColorScheme();

  return (overrideColorScheme ?? colorScheme) === 'dark';
}

export function useColorSchemeTheme(overrideColorScheme?: ColorSchemeName) {
  return useIsDarkColorScheme(overrideColorScheme) ? darkTheme : lightTheme;
}
