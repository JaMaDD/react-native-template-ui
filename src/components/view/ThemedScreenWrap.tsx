import type { FC } from 'react';
import type { ViewStyle } from 'react-native';
import { useInsetsStyle } from '../../hooks/style';
import type { ThemedScreenWrapProps } from '../../types/view';
import { composeStyles } from '../../utils/style/func';
import ThemedView from './ThemedView';

const ThemedScreenWrap: FC<ThemedScreenWrapProps> = ({
  style: themedScreenWrapStyle,
  insets,
  insetTop,
  insetBottom,
  insetLeft,
  insetRight,
  insetsPadding,
  insetPaddingTop,
  insetPaddingBottom,
  insetPaddingLeft,
  insetPaddingRight,
  ...props
}) => {
  const insetsStyle = useInsetsStyle({
    insets,
    insetTop,
    insetBottom,
    insetLeft,
    insetRight,
    insetsPadding,
    insetPaddingTop,
    insetPaddingBottom,
    insetPaddingLeft,
    insetPaddingRight,
  });

  const style = composeStyles<ViewStyle>(themedScreenWrapStyle, insetsStyle);

  return <ThemedView flex={1} style={style} {...props} />;
};

export default ThemedScreenWrap;
