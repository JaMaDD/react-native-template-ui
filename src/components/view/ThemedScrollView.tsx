import { useRestyle } from '@shopify/restyle';
import type { FC } from 'react';
import { ScrollView } from 'react-native';
import { useInsetsStyle } from '../../hooks/style';
import type { ThemedScrollViewProps } from '../../types/view';
import { composeStyles } from '../../utils/style/func';
import { themedScrollViewRestyleFuncs } from '../../utils/theme/restyle';

const ThemedScrollView: FC<ThemedScrollViewProps> = ({
  insets,
  insetTop,
  insetBottom = true,
  insetLeft,
  insetRight,
  insetsPadding,
  insetPaddingTop,
  insetPaddingBottom = 'm',
  insetPaddingLeft,
  insetPaddingRight,
  ...props
}) => {
  const { contentContainerStyle: restyleContentContainerStyle, ...restyles } =
    useRestyle(themedScrollViewRestyleFuncs, props);
  const insetsStyle = useInsetsStyle({
    insets,
    insetTop,
    insetBottom,
    insetLeft,
    insetRight,
    insetsPadding,
    insetPaddingTop,
    insetPaddingBottom,
    insetPaddingLeft,
    insetPaddingRight,
  });

  const contentContainerStyle = composeStyles(
    restyleContentContainerStyle,
    insetsStyle
  );

  return (
    <ScrollView
      showsVerticalScrollIndicator={__DEV__}
      showsHorizontalScrollIndicator={__DEV__}
      contentContainerStyle={contentContainerStyle}
      accessible={true}
      role={'group'}
      {...restyles}
    />
  );
};

export default ThemedScrollView;
