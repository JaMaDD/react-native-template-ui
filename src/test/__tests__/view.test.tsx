import AnimatedThemedView from '../../components/view/AnimatedThemedView';
import ThemedScreenWrap from '../../components/view/ThemedScreenWrap';
import ThemedScrollView from '../../components/view/ThemedScrollView';
import ThemedView from '../../components/view/ThemedView';
import { BorderSize } from '../../utils/theme/const';
import { getThemeColors, getThemeSpacing } from '../utils/func';
import { renderAsync, screen } from '../utils/testingLib';

function getView() {
  return screen.getByRole('group');
}

describe('ThemedView', () => {
  test('Render & Snapshot', async () => {
    const { toJSON } = await renderAsync(<ThemedView />);
    expect(getView()).toBeOnTheScreen();
    expect(toJSON()).toMatchSnapshot();
  });

  test('Change Styles', async () => {
    const { border, backgroundOverlay } = await getThemeColors();
    await renderAsync(
      <ThemedView
        borderWidth={BorderSize.S}
        borderColor={'border'}
        backgroundColor={'backgroundOverlay'}
      />
    );
    expect(getView()).toHaveStyle({
      borderWidth: BorderSize.S,
      borderColor: border,
      backgroundColor: backgroundOverlay,
    });
  });
});

describe('ThemedScreenWrap', () => {
  test('Render & Snapshot', async () => {
    const { toJSON } = await renderAsync(<ThemedScreenWrap />);
    expect(getView()).toBeOnTheScreen();
    expect(toJSON()).toMatchSnapshot();
  });

  test('Change Styles', async () => {
    const { border, backgroundOverlay } = await getThemeColors();
    await renderAsync(
      <ThemedScreenWrap
        borderWidth={BorderSize.S}
        borderColor={'border'}
        backgroundColor={'backgroundOverlay'}
      />
    );
    expect(getView()).toHaveStyle({
      borderWidth: BorderSize.S,
      borderColor: border,
      backgroundColor: backgroundOverlay,
    });
  });

  test('Change Insets', async () => {
    const { s, l } = await getThemeSpacing();
    await renderAsync(
      <ThemedScreenWrap
        insets={true}
        insetPaddingLeft={'s'}
        insetPaddingRight={'l'}
      />
    );
    expect(getView()).toHaveStyle({ paddingLeft: s, paddingRight: l });
  });
});

describe('ThemedScrollView', () => {
  test('Render & Snapshot', async () => {
    const { toJSON } = await renderAsync(<ThemedScrollView />);
    expect(getView()).toBeOnTheScreen();
    expect(toJSON()).toMatchSnapshot();
  });

  test('Change Insets', async () => {
    const { xxs, xxl } = await getThemeSpacing();
    await renderAsync(
      <ThemedScrollView
        insetTop={true}
        insetBottom={true}
        insetPaddingTop={'xxl'}
        insetPaddingBottom={'xxs'}
      />
    );
    expect(getView()).toHaveProp('contentContainerStyle', {
      paddingTop: xxl,
      paddingBottom: xxs,
    });
  });
});

describe('AnimatedThemedView', () => {
  test('Render & Snapshot', async () => {
    const { toJSON } = await renderAsync(<AnimatedThemedView />);
    expect(getView()).toBeOnTheScreen();
    expect(toJSON()).toMatchSnapshot();
  });

  test('Change Styles', async () => {
    const { border, backgroundOverlay } = await getThemeColors();
    await renderAsync(
      <AnimatedThemedView
        borderWidth={BorderSize.S}
        borderColor={'border'}
        backgroundColor={'backgroundOverlay'}
      />
    );
    expect(getView()).toHaveStyle({
      borderWidth: BorderSize.S,
      borderColor: border,
      backgroundColor: backgroundOverlay,
    });
  });
});
