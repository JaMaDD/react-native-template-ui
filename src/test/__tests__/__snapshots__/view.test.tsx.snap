// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedThemedView Render & Snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    jestAnimatedProps={
      {
        "value": {},
      }
    }
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    nativeID="0"
    role="group"
    style={
      [
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  />
</View>
`;

exports[`ThemedScreenWrap Render & Snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    role="group"
    style={
      [
        {
          "backgroundColor": "#ffffff",
          "flex": 1,
        },
        {},
      ]
    }
  />
</View>
`;

exports[`ThemedScrollView Render & Snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    accessible={true}
    contentContainerStyle={
      {
        "paddingBottom": 16,
      }
    }
    role="group"
    showsHorizontalScrollIndicator={true}
    showsVerticalScrollIndicator={true}
    style={
      [
        {},
      ]
    }
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`ThemedView Render & Snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    role="group"
    style={
      [
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  />
</View>
`;
